keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,299 @@
+
+- Volume limits does not yet integrate with cluster autoscaler if all nodes in the cluster are running at maximum volume limits.
+
+## Proposal
+
+* Track volume limits for CSI driver in `CSINode` objects instead of `Node`. Limit in `CSINode` is used instead of limit coming from `Node` object whenever available for same in-tree volume type. This mean scheduler will always try to translate in-tree driver name to CSI driver name whenever `CSINode` object has same in-tree volume type (even if migration is off).
+  * To get rid of prefix + SHA for `ResourceName` of CSI volumes.
+  * So in-tree volume plugin can share limits with CSI driver that uses the same storage backend.
+
+### API Change
+
+CSINode is split into `spec` and `status`. `spec` contains list of drivers installed to the node and their properties that do not change during lifetime of a driver. `status` is missing right now, but it may be used later e.g. for driver health that changes in time.
+
+We expect that limits of a CSI driver does not change during lifetime of a driver and therefore we put the resource limits into `CSINodeSpec`. The only way for a driver to change the limits is to deregister and register again, e.g. by restarting its container.
+
+```go
+// Until further notice, this is existing API to introduce full context.
+
+type CSINode struct {
+	...
+
+	// spec is the specification of CSINode
+	Spec CSINodeSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`}
+}
+
+// CSINodeSpec holds information about the specification of all CSI drivers installed on a node
+type CSINodeSpec struct {
+	// drivers is a list of information of all CSI Drivers existing on a node.
+	// If all drivers in the list are uninstalled, this can become empty.
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Drivers []CSINodeDriver `json:"drivers" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=drivers"`
+}
+
+// CSINodeDriver holds information about the specification of one CSI driver installed on a node
+type CSINodeDriver struct {
+    // ...
+
+    // NEW API STARTS HERE
+    // Allocatable represents the resources of a node that are available for scheduling for volumes of this driver.
+    Allocatable VolumeLimits
+}
+
+// VolumeLimits is a set of resource limits for scheduling of volumes.
+type VolumeLimits struct {
+	// Volumes is maximum number of volumes provided by the CSI driver that can be used by the node
+	// "nil" represents no limits - the node can handle any number of volumes of the driver.
+	Volumes *int32 `json:"volumes,omitempty" protobuf:"varint,1,opt,name=volumes"`
```

Comment:
What about using "Count" instead of Volumes, since Volume is already in the name of the struct.

Then in the future, if we add capacity, then the new field would be called Capacity

Reply:
yeah I like count more. fixed